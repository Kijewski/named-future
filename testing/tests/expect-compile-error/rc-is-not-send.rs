use std::rc::Rc;

use futures_lite::future::block_on;
use named_future::named_future;

#[named_future(Send)]
async fn async_drop(counter: Rc<()>) {
    drop(counter);
}

fn main() {
    block_on(async_drop(Rc::new(())));
}
