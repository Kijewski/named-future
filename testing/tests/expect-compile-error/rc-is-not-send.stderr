error: future cannot be sent between threads safely
 --> tests/expect-compile-error/rc-is-not-send.rs:6:16
  |
6 |   #[named_future(Send)]
  |  ________________^
7 | | async fn async_drop(counter: Rc<()>) {
  | |___________________^ future returned by `async_drop` is not `Send`
  |
  = help: within `impl Future<Output = ()>`, the trait `Send` is not implemented for `Rc<()>`
note: future is not `Send` as this value is used across an await
 --> tests/expect-compile-error/rc-is-not-send.rs:7:10
  |
7 | async fn async_drop(counter: Rc<()>) {
  |          ^^^^^^^^^- ------- has type `Rc<()>` which is not `Send`
  |          |        |
  |          |        `counter` is later dropped here
  |          await occurs here, with `counter` maybe used later
note: required by a bound in `ensure_send`
 --> $WORKSPACE/src/machinery.rs
  |
  |     Fut: marker::Send,
  |          ^^^^^^^^^^^^ required by this bound in `ensure_send`
